docs/ossm/ossm2-migration/cluster-wide/README.md

```diff
@@ -0,0 +1,165 @@
+    istio-system   install-istio-system   6/6     ComponentsReady   ["default"]   2.6.4     115m
+    ```
+1. Prepare the `Istio` resource yaml named `ossm-3.yaml` to be deployed to the same namespace as the 2.6 control plane:
+
+    Here we are not using any `discoverySelectors` so the control plane will have access to all namespaces. In case you want to define `discoverySelectors`, keep in mind that all data plane namespaces you are planning to migrate from 2.6 must be matched.
+
+    ```yaml
+   apiVersion: sailoperator.io/v1alpha1
+   kind: Istio
+   metadata:
+     name: ossm-3
+   spec:
+     namespace: istio-system # the same namespace where we run the 2.6 control plane
+     version: v1.24.1
+   ```
+1. Apply the `Istio` resource yaml:
+
+    ```sh
+    oc apply -f ossm-3.yaml
+    ```
+1. Verify that new `istiod` is using existing root certificate:
+
+    ```sh
+    oc logs deployments/istiod-ossm-3 -n istio-system | grep 'Load signing key and cert from existing secret'
+    2024-12-18T08:13:53.788959Z	info	pkica	Load signing key and cert from existing secret istio-system/istio-ca-secret
+    ```
+
+> [!CAUTION]
+> If you use the name `default` for your 3.0 `Istio` resource with the `InPlace` upgrade strategy or you have configured the `default` revision tag, the 3.0 control plane will try to inject side cars to all pods in namespaces with the `istio-injection=enabled` label and all pods with the `sidecar.istio.io/inject="true"` label after next restart of the workloads.
+
+#### Migrate Workloads
+This guide will use `istio.io/rev=ossm-3` label. In case you are using the `default` revision or you have configured the `default` revision tag, it's possible to use `istio-injection=enabled` label. The following steps will work with that label (instead of `istio.io/rev=ossm-3`) too.
+
+1. Update injection labels on the data plane namespace
+
+    Here we're adding two labels to the namespace:
+
+    1. The `istio.io/rev=ossm-3` label which ensures that any new pods that get created in that namespace will connect to the 3.0 proxy. In our example, the 3.0 revision is named `ossm-3`
+    2. The `maistra.io/ignore-namespace: "true"` label which will disable sidecar injection for 2.6 proxies in the namespace. This ensures that 2.6 will stop injecting proxies in this namespace and any new proxies will be injected by the 3.0 control plane. Without this, the 2.6 injection webhook will try to inject the pod and it will connect to the 2.6 proxy as well as refuse to start since it will have the 2.6 cni annotation.
+
+        > **_NOTE:_** that once you apply the `maistra.io/ignore-namespace` label, any new pod that gets created in the namespace will be connected to the 3.0 proxy. Workloads will still be able to communicate with each other though regardless of which control plane they are connected to.
+
+    ```sh
+    oc label ns bookinfo istio.io/rev=ossm-3 maistra.io/ignore-namespace="true" --overwrite=true
```

Comment:
> maistra.io/ignore-namespace="true"

I don't think you need this label. Clusterwide does not have the same problem as MultiTenant where the istiod cannot do discovery in the namespace if the namespace is removed from the SMMR because ClusterWide always does discovery in all namespaces (unless you limit with discovery selectors) regardless of what is in the SMMR.

You _do_ need to remove the `istio-injection: enabled` label though if you are using this for 2.6 and you are using `istio.io/rev` for 3.0 since `istio-injection: enabled` seems to take precedence over `istio.io/rev`. For cluster wide, we can remove the namespace from the SMMR and the proxies will continue to communicate with one another. I'm not sure which scenario we want to base the example off of but here's an updated example that adds the `istio.io/rev` label and removes the `istio-injection` label which effectively removes the namespace from the SMMR.

```suggestion
    oc label ns bookinfo istio.io/rev=ossm-3 istio-injection- --overwrite=true
```

Reply:
Right, I also thought about this option. It will work fine for scenarios without `istio-injection` label but it would complicate a lot scenarios where customers are using `istio-injection` label in 2.6 and want to continue using it for 3.0. 
Reasons I decided to use `maistra.io/ignore-namespace="true"`:

- to avoid at least 3 different migration scenarios for customers who a) have `istio-injection` label in 2.6 and want to continue using it for 3.0, b) have `istio-injection` label in 2.6 and don't want to use it in 3.0 and c) customers without `istio-injection` label in 2.6.
- sceario a) would be much more complicated compared to what we have now with `maistra.io/ignore-namespace="true"` e.g. changing `spec.memberSelectors` in SMMR, re-labeling namespaces. I have not tested this scenario without `maistra.io/ignore-namespace="true"` so I'm not even sure if it's possible.
- using `maistra.io/ignore-namespace="true"` will work for all 3 scenarios
- multitenant and clusterwide migration steps will be similar

Reply:
> a) have istio-injection label in 2.6 and want to continue using it for 3.0

The default for cluster wide seems to use `istio-injection` so I'm guessing this will be the most common scenario.

> b) have istio-injection label in 2.6 and don't want to use it in 3.0

In this case you have to remove `istio-injection` before the 3.0 controlplane will inject a sidecar. Even if the `maistra.io/ignore-namespace` label is present and the 2.6 controlplane doesn't inject, if the `istio-injection` label is present then the revisioned 3.0 controlplane won't inject the sidecar either.

> a) would be much more complicated compared to what we have now with maistra.io/ignore-namespace="true" e.g. changing spec.memberSelectors in SMMR, re-labeling namespaces. I have not tested this scenario without maistra.io/ignore-namespace="true" so I'm not even sure if it's possible.

I agree I think you'll need `maistra.io/ignore-namepsace` here so that only the 3.0 controlplane does injection.

> c) customers without istio-injection label in 2.6.

This is the easiest scenario since it's just a normal canary upgrade at that point and all you need to do is update `istio.io/rev`.

Reply:
I don't think one approach will work for all 3 scenarios because you'll need to perform different steps for each.

>  b) have istio-injection label in 2.6 and don't want to use it in 3.0 

For this scenario you'll have to remove the `istio-injection` label regardless of whether or not you use `maistra.io/ignore-namespace`. Whereas for scenario a) you don't want to remove `istio-injection`.

Reply:
To be sure, I tested scenario b) where my namespace is this:
```
apiVersion: v1
kind: Namespace
metadata:
  labels:
    istio-injection: enabled
    istio.io/rev: ossm-3
    maistra.io/ignore-namespace: "true"
    maistra.io/member-of: istio-system
  name: bookinfo
```
and I restarted my `reviews-v2` workload and now the pod has no sidecar at all:
```
reviews-v2-69c6fcfd69-qjb2p       1/1     Running   0          22s
```
because with `maistra.io/ignore-namespace: "true"` 2.6 won't inject and since the `istio-injection: enabled` label is present the revisioned 3.0 controlplane won't inject either.

Reply:
Ah, I know what you mean now. I thought that removal of `istio-injection`  label in case b) would be implicit. As switching from `istio-injection` to `istio.io/rev` will also include the removal. Current doc is showing scenario c) which is probably not the best one to show. Thank you.